projects/plugins/jetpack/modules/subscriptions.php

```diff
@@ -989,6 +992,189 @@ public function register_post_meta() {
+			'email_subscribers' => isset( $subscriber_data['email_subscribers'] ) ? (int) $subscriber_data['email_subscribers'] : 0,
+			'paid_subscribers'  => isset( $subscriber_data['paid_subscribers'] ) ? (int) $subscriber_data['paid_subscribers'] : 0,
+			'all_subscribers'   => isset( $subscriber_data['all_subscribers'] ) ? (int) $subscriber_data['all_subscribers'] : 0,
+			'subscriber_list'   => isset( $subscriber_data['subscriber_list'] ) && is_array( $subscriber_data['subscriber_list'] ) ? $subscriber_data['subscriber_list'] : array(),
+		);
+
+		update_post_meta( $post_ID, '_jetpack_newsletter_subscribers_when_sent', $data_to_store );
+	}
+
+	/**
+	 * Get subscriber data from WordPress.com API.
+	 *
+	 * @since $$next-version$$
+	 *
+	 * @return array Subscriber data including counts and email list.
+	 */
+	private function get_subscriber_data() {
+		$subscriber_data = array(
+			'email_subscribers' => 0,
+			'paid_subscribers'  => 0,
+			'all_subscribers'   => 0,
+			'subscriber_list'   => array(),
+		);
+
+		// Only fetch if Jetpack is connected.
+		if ( ! Jetpack::is_connection_ready() ) {
+			return $subscriber_data;
+		}
+
+		$site_id = Jetpack_Options::get_option( 'id' );
+
+		// First, get subscriber counts from stats endpoint.
+		$stats_path     = sprintf( '/sites/%d/subscribers/stats', $site_id );
+		$stats_response = Client::wpcom_json_api_request_as_blog(
+			$stats_path,
+			'2',
+			array(),
+			null,
+			'wpcom'
+		);
+
+		if ( ! is_wp_error( $stats_response ) ) {
+			$stats_code = wp_remote_retrieve_response_code( $stats_response );
+			if ( 200 === $stats_code ) {
+				$subscriber_counts = json_decode( wp_remote_retrieve_body( $stats_response ), true );
+				if ( is_array( $subscriber_counts ) ) {
+					if ( isset( $subscriber_counts['counts']['email_subscribers'] ) ) {
+						$subscriber_data['email_subscribers'] = (int) $subscriber_counts['counts']['email_subscribers'];
+					}
+					if ( isset( $subscriber_counts['counts']['paid_subscribers'] ) ) {
+						$subscriber_data['paid_subscribers'] = (int) $subscriber_counts['counts']['paid_subscribers'];
+					}
+					if ( isset( $subscriber_counts['counts']['all_subscribers'] ) ) {
+						$subscriber_data['all_subscribers'] = (int) $subscriber_counts['counts']['all_subscribers'];
+					}
+				}
+			}
+		}
+
+		// Fetch the actual subscriber list with emails.
+		$subscriber_emails                  = $this->fetch_all_subscribers( $site_id );
+		$subscriber_data['subscriber_list'] = $subscriber_emails;
+
+		return $subscriber_data;
+	}
+
+	/**
+	 * Fetch all subscribers from WordPress.com API with pagination.
+	 *
+	 * @since $$next-version$$
+	 *
+	 * @param int $site_id Site ID.
+	 * @return array Array of subscriber data, each containing 'email' and 'is_paid' keys.
+	 */
+	private function fetch_all_subscribers( $site_id ) {
+		$subscriber_emails = array();
+		$page              = 1;
+		$per_page          = 100; // Maximum per page to minimize requests.
+
+		while ( true ) {
```

Comment:
Since we're potentially making multiple calls to WordPress.com to get all that data, I think this should be saved in a transient, to save any too frequent calls.

Reply:
I added some transients, they are currently set at 5 minutes.